### PR TITLE
fix (refs T30474): reduce inconsistent code duplication

### PR DESCRIPTION
**Pull request by dresslerdemos**
Enhance the `OwnsProcedureConditionFactory` implementation, so that it can not only create conditions to match unknown users against a known procedure, but also to match unknown procedures against a known user. Logically the comparison is the same in both cases, but their implementation differs significantly.

Keeping the implementations as close together as possible avoids divergences, like were already present in
`ProcedureService::getAdminProcedureConditions`. The matching logic in this method behaved very similar to the one executed via `ProcedureAccessEvaluator::isOwningProcedure` (which uses `OwnsProcedureConditionFactory`), but was not exactly the same, causing a bug. Hence, `isOwningProcedure` was selected as "truth" and the logic in `getAdminProcedureConditions` was dropped.

**Ticket:** https://yaits.demos-deutschland.de/T30474

### PR Checklist

Delete the checkbox if it doesn't apply/isn't necessary.

- [ ] Tests updated/created
- [ ] Update documentation
- [ ] Link all relevant tickets
- [ ] Move the tickets on the board accordingly

### CODE REVIEW

File: demosplan/DemosPlanCoreBundle/Logic/OwnsProcedureConditionFactory.php [hunk 1]
```diff
@@ -13,120 +13,154 @@
...
-                $userIsAuthorized,
-                $planningAgencyIsAuthorized
-            );
-        }
-
-        return $orgaOwnsProcedure;
+        // T8427: allow access by manually configured users if the config is set to `true`,
+        // overwriting the organisation-based access
+        return $this->globalConfig->hasProcedureUserRestrictedAccess()
+            ? $this->userIsExplicitlyAuthorized()
```
Comment (mussbach) [inline demosplan/DemosPlanCoreBundle/Logic/OwnsProcedureConditionFactory.php:97]: Isnt in the true case here now the permissions for planningAgencies ditched? Now only authorized users are checked, not anymore whether planning agency is permitted to procedure, if I see it correctly

File: demosplan/DemosPlanCoreBundle/Logic/OwnsProcedureConditionFactory.php [hunk 2]
```diff
@@ -13,120 +13,154 @@
...
+            : $this->userOwnsProcedureViaOrgaOfUserThatCreatedTheProcedure();
     }
 
     /**
-     * Returns a condition to match users having the roles to theoretically own a procedure.
+     * Returns a condition to match users having the roles in the given customer to theoretically own a procedure.
      *
-     * @return FunctionInterface<bool>
+     * @return list<FunctionInterface<bool>>
```
Comment (mussbach) [inline demosplan/DemosPlanCoreBundle/Logic/OwnsProcedureConditionFactory.php:104]: is this correct?

File: demosplan/DemosPlanCoreBundle/Logic/OwnsProcedureConditionFactory.php [hunk 3]
```diff
@@ -13,120 +13,154 @@
...
-                    Role::HEARING_AUTHORITY_ADMIN,
-                    Role::HEARING_AUTHORITY_WORKER,
-                ],
-                ['roleInCustomers', 'role', 'code']
-            );
+            $ownsOrgaRoleCondition = [
+                $this->conditionFactory->propertyHasAnyOfValues($relevantRoles, ['roleInCustomers', 'role', 'code']),
+                $this->isUserInCustomer($customer),
+            ];
+        } else {
```
Comment (mussbach) [inline demosplan/DemosPlanCoreBundle/Logic/OwnsProcedureConditionFactory.php:130]: You can avoid this else by setting this as default value of the variable

File: demosplan/DemosPlanCoreBundle/Logic/OwnsProcedureConditionFactory.php [hunk 4]
```diff
@@ -13,120 +13,154 @@
...
             // ist es ein PLanungsbüro?
-            $planningAgencyOwnsProcedure = $this->conditionFactory->propertyHasValue(
-                Role::PRIVATE_PLANNING_AGENCY,
-                ['roleInCustomers', 'role', 'code']
-            );
+            $planningAgencyOwnsProcedure = [
+                $this->conditionFactory->propertyHasValue($relevantRole, ['roleInCustomers', 'role', 'code']),
+                $this->isUserInCustomer($customer),
+            ];
+        } else {
```
Comment (mussbach) [inline demosplan/DemosPlanCoreBundle/Logic/OwnsProcedureConditionFactory.php:162]: you can avoid this else

File: demosplan/DemosPlanCoreBundle/Logic/ProcedureAccessEvaluator.php [hunk 5]
```diff
@@ -87,51 +100,61 @@ public function isOwningProcedure(User $user, Procedure $procedure): bool
     }
 
     /**
+     * Compiles a list of conditions to check if a procedure or procedure template is owned by the given user.
+     *
+     * A procedure is owned a user if all conditions in the returned list match the procedure/user respectively.
```
Comment (mussbach) [inline demosplan/DemosPlanCoreBundle/Logic/ProcedureAccessEvaluator.php:105]:
```suggestion
     * A procedure is owned by a user if all conditions in the returned list match the procedure/user respectively.
```

File: demosplan/DemosPlanCoreBundle/Logic/OwnsProcedureConditionFactory.php [hunk 6]
```diff
@@ -135,11 +169,91 @@ public function hasPlanningAgencyRole(): FunctionInterface
...
+
+        return $this->conditionFactory->propertyHasValue($procedure->getOrgaId(), ['orga', 'id']);
+    }
+
+    /**
+     * @return ClauseFunctionInterface<bool>
+     *
+     * @throws PathException
+     */
+    public function isUndeletedProcedure(): ClauseFunctionInterface
```
Comment (mussbach) [inline demosplan/DemosPlanCoreBundle/Logic/OwnsProcedureConditionFactory.php:247]: Undeleted sounds a bit strange here. Perhaps `isNotDeleted`?

**Review by mussbach — CHANGES_REQUESTED** [review]
Zwei Herzen ach.. I absolutely see where you come from and this might well be a good Idea but it is quite uncommon to pass two such different Objects into the check. In this case this might be a very good possibility and let's try it, but I do not dare to predict how we look at this in one or two years.
Some remarks, but basically fine though it is very hard to review

File: demosplan/DemosPlanCoreBundle/Logic/OwnsProcedureConditionFactory.php (hunk 1 above)
Comment (dresslerdemos) [reply demosplan/DemosPlanCoreBundle/Logic/OwnsProcedureConditionFactory.php:97]: Yes that was one of the weird inconsistencies that was noticed some time (about a year?) ago but not dared to be touched. I tried to confirm the intended behavior with Andrea via Slack. The "new" behavior (it was already implemented in a different place) will still authorize users via their planning agency (`$this->isAuthorizedViaPlanningAgency();`), but
1. to be authorized, users must have the `PRIVATE_PLANNING_AGENCY` role
2. the authorization via planning agency is active, regardless of the `hasProcedureUserRestrictedAccess` global config setting.

Again this is not really new, but consolidated behavior and so far makes sense to me logically.

How planning agency are authorized can be found by searching for the calls of `isAuthorizedViaPlanningAgency`.

File: demosplan/DemosPlanCoreBundle/Logic/OwnsProcedureConditionFactory.php (hunk 2 above)
Comment (dresslerdemos) [reply demosplan/DemosPlanCoreBundle/Logic/OwnsProcedureConditionFactory.php:104]: `list` is a type introduced by phpstan and as far as I know accepted by PHPStorm. It denotes an array of values whose keys are integers starting with `0` that increase without holes.

`FunctionInterface<bool>` is "good enough". There are inconsistencies when to use `FunctionInterface` vs `ClauseFunctionInterface` (the latter extending the former). There is a clear solution for that, but it requires some refactoring which execution needs to be planned and approved first.

Comment (dresslerdemos) [reply demosplan/DemosPlanCoreBundle/Logic/OwnsProcedureConditionFactory.php:104]: `list` is a type introduced by phpstan and as far as I know accepted by PHPStorm. It denotes an array of values whose keys are integers starting with `0` that increase without holes.

`FunctionInterface<bool>` is "good enough". There are inconsistencies when to use `FunctionInterface` vs `ClauseFunctionInterface` (the latter extending the former). There is a clear solution for that, but it requires some refactoring which execution needs to be planned and approved first.

File: demosplan/DemosPlanCoreBundle/Logic/OwnsProcedureConditionFactory.php (hunk 3 above)
Comment (dresslerdemos) [reply demosplan/DemosPlanCoreBundle/Logic/OwnsProcedureConditionFactory.php:130]: I'm fine with that for simple cases like `null` or primitive types as default, but here a method is called and an object would be created, which is instantly thrown away if the `if` branch is entered. This does not seem like a good design choice, as a programmer not familiar with this code may feel inclined to check for side effects in the logic of `$this->conditionFactory->false()` even though he is interested in the `if` case only.

File: demosplan/DemosPlanCoreBundle/Logic/OwnsProcedureConditionFactory.php (hunk 4 above)
Comment (dresslerdemos) [reply demosplan/DemosPlanCoreBundle/Logic/OwnsProcedureConditionFactory.php:162]: https://github.com/demos-europe/demosplan-core/pull/2662/files#r1478637156

File: demosplan/DemosPlanCoreBundle/Logic/OwnsProcedureConditionFactory.php (hunk 6 above)
Comment (dresslerdemos) [reply demosplan/DemosPlanCoreBundle/Logic/OwnsProcedureConditionFactory.php]: I renamed it to `isNotDeletedProcedure`, to keep the clarity that this checks the deletion of the procedure, not the user. e54af829944d655083c2a208ead4fcc931ff6bbd

**Comment by dresslerdemos** [comment]
> quite uncommon to pass two such different Objects into the check

My initial idea was to use inheritance from an common `OwnsProcedureConditionFactoryInterface` and have one implementation taking a `User` and another implementation taking a `Procedure`. But even though this would have added a connection via the same parent methods in the interface, I reckoned the likelyhood for changes in only one class was too great (leave alone the complexity of three instead of one class).

One thing that could happen in the future would be that both procedure and user are variable, to fetch all users from the database that own *any* procedure in the current customer. This would introduce a third case in each method, `$userOrProcedure` being `null`.

**Review by mussbach — APPROVED** [review]
concerns are resolved, lgtm, code review only

**Pull request merged**